### PR TITLE
Make blob migrations more resiliant/scalable

### PR DESCRIPTION
**Pull request by millerdev**
- Add support for stop/resume migration
- Add support for retry on documents with concurrent modifications.
- Improve progresss output during migration.
- Do not use `<DocType>.wrap(doc)`, which can be problematic for very old documents.

@dannyroberts @snopoke cc @benrudolph

### CODE REVIEW

File: corehq/blobs/migrate.py [hunk 1]
```diff
@@ -259,45 +275,81 @@ def encode_content(data):
...
+                            "content_type": meta["content_type"],
+                            "content": encode_content(meta["content"]),
+                        }
+                        for name, meta in doc["_attachments"].items()
+                    }
+                else:
+                    backup_doc = doc
+                f.write('{}\n'.format(json.dumps(backup_doc)))
+                f.flush()
+                ok = doc_migrator.migrate(doc, couchdb)
```
Comment (millerdev) [inline corehq/blobs/migrate.py]: Documents are now migrated as they are loaded from the database rather than loading all documents into a file and then migrating the file in a separate phase. This has several advantages:

- The migration can be stopped and resumed without incurring a massive reload of unmigrated documents.
- If the migration log gets too big, the migration can be stopped to allow space to be freed up before resuming the migration.
- Less chance of concurrent modification errors because it is a relatively short time from when a document is loaded from couch to when it is migrated and finally saved.

As before, documents, including all attachments, are serialized to a migration log before they are modified as a safety precaution in case something goes wrong with the migration.

File: corehq/blobs/migrate.py [hunk 2]
```diff
@@ -127,122 +123,142 @@
...
+    def migrate(self, doc, couchdb):
+        attachments = doc.pop("_attachments")
+        external_blobs = doc.setdefault("external_blobs", {})
+        obj = BlobHelper(doc, couchdb)
+        try:
+            with obj.atomic_blobs():
+                for name, data in list(attachments.iteritems()):
+                    if name in external_blobs:
+                        continue  # skip attachment already in blob db
+                    obj.put_attachment(name=name, **data)
```
Comment (snopoke) [inline corehq/blobs/migrate.py]: Will this do a `save` on the doc even if no changes have been made?

File: corehq/util/couch_helpers.py [hunk 3]
```diff
@@ -169,6 +171,153 @@ def paginate_view(db, view_name, chunk_size,
...
+    """
+
+    def __init__(self, db, doc_types, iteration_key, chunk_size=100):
+        if isinstance(doc_types, str):
+            raise TypeError("expected list of strings, got %r" % (doc_types,))
+        self.db = db
+        self.original_doc_types = doc_types = sorted(doc_types)
+        self.iteration_key = iteration_key
+        self.chunk_size = chunk_size
+        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
```
Comment (snopoke) [inline corehq/util/couch_helpers.py]: Should this use the sorted list of doc types?

File: corehq/util/couch_helpers.py [hunk 4]
```diff
@@ -169,6 +171,153 @@ def paginate_view(db, view_name, chunk_size,
...
+        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
+        try:
+            self.state = db.get(self.iteration_id)
```
Comment (snopoke) [inline corehq/util/couch_helpers.py]: The state docs are saved in whichever DB we're migrating docs in. Wondering if it wouldn't be better to always save them in the same DB (probably `meta`)

File: corehq/util/couch_helpers.py [hunk 5]
```diff
@@ -169,6 +171,153 @@ def paginate_view(db, view_name, chunk_size,
...
+            # new iteration
+            self.state = {
+                "_id": self.iteration_id,
+                "doc_type": "ResumableDocsByTypeIteratorState",
+                "retry": {},
+
+                # for humans
+                "name": iteration_name,
+                "timestamp": datetime.datetime.utcnow().isoformat()
+            }
```
Comment (snopoke) [inline corehq/util/couch_helpers.py]: Any reason not to use a JsonObject for the `ResumableDocsByTypeIteratorState` rather than a raw dict?

File: corehq/util/couch_helpers.py (hunk 4 above)
Comment (snopoke) [inline corehq/util/couch_helpers.py]: Not a huge fan of having DB access in `__init__` functions but probably OK given the use case for this class.

**Comment by snopoke** [comment]
A few minor comments but looks good. The retry logic is a bit hard to follow but tests looks good.

File: corehq/blobs/migrate.py (hunk 2 above)
Comment (millerdev) [inline corehq/blobs/migrate.py]: yes, but there should always be a change since there are always attachments to migrate when this function is called.

File: corehq/util/couch_helpers.py (hunk 3 above)
Comment (millerdev) [inline corehq/util/couch_helpers.py]: yep, it does. see line 197

File: corehq/util/couch_helpers.py (hunk 4 above)
Comment (millerdev) [inline corehq/util/couch_helpers.py]: Where would you put this initialization code if not in `__init__`?

File: corehq/util/couch_helpers.py (hunk 5 above)
Comment (millerdev) [inline corehq/util/couch_helpers.py]: I had considered it, but I decided to do this to minimize dependencies. I don't think using `JsonObject` would reduce the amount of code much if at all.

File: corehq/util/tests/test_couch_helpers.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    def tearDownClass(cls):
+        for doc_id in set(d["_id"] for d in cls.docs):
+            try:
+                cls.db.delete_doc(doc_id)
+            except ResourceNotFound:
+                pass
+
+    def setUp(self):
+        self.domain = "TEST"
+        self.sorted_keys = ["{}-{}".format(n, i)
```
Comment (millerdev) [inline corehq/util/tests/test_couch_helpers.py]: note to self: `self.domain` is not used. `self.sorted_keys = ...` could be moved into `setUpClass`

File: corehq/util/couch_helpers.py [hunk 7]
```diff
@@ -169,6 +171,153 @@ def paginate_view(db, view_name, chunk_size,
...
+                retried[doc_id] = retries
+                try:
+                    yield self.db.get(doc_id)
+                except ResourceNotFound:
+                    pass
+
+        # save iteration state without offset to signal completion
+        self.state.pop("offset", None)
+        self.state["retry"] = {}
+        self._save_state()
```
Comment (millerdev) [inline corehq/util/couch_helpers.py]: note to self: should save state without offset before retry logic, and only save again if any docs were retried.

File: corehq/util/couch_helpers.py (hunk 5 above)
Comment (snopoke) [inline corehq/util/couch_helpers.py]: I wasn't thinking about reducing code. Using JsonObject would make it clear
what the schema for this dict is and make working with it easier.

On 27 May 2016 at 14:29, Daniel Miller <notifications@github.com> wrote:

> In corehq/util/couch_helpers.py
> <https://github.com/dimagi/commcare-hq/pull/11792#discussion_r64897225>:
>
> > +        self.chunk_size = chunk_size
> > +        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
> > +        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
> > +        try:
> > +            self.state = db.get(self.iteration_id)
> > +        except ResourceNotFound:
> > +            # new iteration
> > +            self.state = {
> > +                "_id": self.iteration_id,
> > +                "doc_type": "ResumableDocsByTypeIteratorState",
> > +                "retry": {},
> > +
> > +                # for humans
> > +                "name": iteration_name,
> > +                "timestamp": datetime.datetime.utcnow().isoformat()
> > +            }
>
> I had considered it, but I decided to do this to minimize dependencies. I
> don't think using JsonObject would reduce the amount of code much if at
> all.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/11792/files/c110713a9e9c98d7ac6793275af103f2941f1987#r64897225>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAPPBq-DBLHubl3b5FCco195JHHtoEmSks5qFuOTgaJpZM4In7ch>
> .
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa
